libs/deck-layers/src/layers/fourwings/FourwingsHeatmapTileLayer.ts

```diff
@@ -73,81 +46,106 @@ export class FourwingsHeatmapTileLayer extends CompositeLayer<
-    // Select only 5% of elements
-    return value && index % 20 === 1
-  }
-
-  calculateColorDomain = (tiles: Tile2DHeader<FourwingsHeatmapTileData>[]) => {
+  calculateColorDomain = (comparisonMode?: HeatmapAnimatedMode) => {
     // TODO use to get the real bin value considering the NO_DATA_VALUE and negatives
     // NO_DATA_VALUE = 0
     // SCALE_VALUE = 0.01
     // OFFSET_VALUE = 0
-    // const fa = performance.now()
-    const currentZoomTiles = tiles.filter(
+    const currentZoomTiles = (this.state.tiles as FourwingsHeatmapTileLayerState['tiles']).filter(
       (tile) => tile.zoom === Math.round(this.context.viewport.zoom)
     )
     if (!currentZoomTiles?.length) {
       return this.getColorDomain()
     }
+    if (comparisonMode === HeatmapAnimatedMode.Bivariate) {
+      let allValues: [number[], number[]] = [[], []]
+      currentZoomTiles.forEach((tile) => {
+        if (!tile.content) return
+        ;(tile.content.length > MAX_VALUES_PER_TILE
+          ? tile.content.filter(filterElementByPercentOfIndex)
+          : tile.content
+        ).forEach((feature) => {
+          feature.properties?.values.forEach((sublayerValues, sublayerIndex) => {
+            allValues[sublayerIndex].push(...sublayerValues.filter(filterElementByPercentOfIndex))
+          })
+        })
+      })
+      if (!allValues.length) {
+        return this.getColorDomain()
+      }
+      const steps = allValues.map((sublayerValues) =>
+        ckmeans(sublayerValues, Math.min(sublayerValues.length, 4)).map((step) => step[0])
+      )
+      return steps
+    }
+
     const allValues = currentZoomTiles.flatMap((tile) => {
       if (!tile.content) return []
       return (
         tile.content.length > MAX_VALUES_PER_TILE
-          ? tile.content.filter(this.filterElementByPercentOfIndex)
+          ? tile.content.filter(filterElementByPercentOfIndex)
           : tile.content
       ).flatMap((feature) =>
         feature.properties?.values.flatMap((values) => {
-          return (values || []).filter(this.filterElementByPercentOfIndex)
+          return (values || []).filter(filterElementByPercentOfIndex)
         })
       )
     })
-    // console.log('allValues:', allValues.length)
-    // const a = performance.now()
     if (!allValues.length) {
       return this.getColorDomain()
     }
     const steps = ckmeans(allValues, Math.min(allValues.length, COLOR_RAMP_DEFAULT_NUM_STEPS)).map(
       (step) => step[0]
     )
-    // const b = performance.now()
-    // console.log('ckmeans time:', b - a)
-    // const fb = performance.now()
-    // console.log('steps:', steps)
-    // console.log('calculateColorDomain time:', fb - fa)
     return steps
   }
 
-  updateColorDomain = debounce((tiles) => {
+  updateColorDomain = debounce(() => {
     requestAnimationFrame(() => {
-      this.setState({ colorDomain: this.calculateColorDomain(tiles) })
+      this.setState({ colorDomain: this.calculateColorDomain(this.state.comparisonMode) })
```

Comment:
I think would be more clear to have all the needed data in the same format, I mean:
- sending all the data as function params so we can extract the function out of the class
- reading all the data directly from the state

but in this case we send the comparisonMode as a parameter but the tiles from the [state](https://github.com/GlobalFishingWatch/frontend/pull/2560/commits/7d3bb738a2d5fb6207838090cfcef96ca4b8c7f0#diff-019339b86e8de09cbdbbb40dc141be52d4e4c5278f335c9451dbcb8a94d6f34aR91) WDYT ?